http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,194 @@
+            )
+            _ <- clusterHelper.stopCluster(resolvedCluster)
+          } yield ()
+        }
+        case Some(resolvedCluster) =>
+          IO.raiseError(
+            new WorkbenchException(
+              s"Failed to process StopUpdateMessage for Cluster ${resolvedCluster.projectNameString}. This is likely due to a mismatch in state between the db and the message, or an improperly formatted machineConfig in the message. Cluster details: ${resolvedCluster}"
+            )
+          )
+        case None =>
+          IO.raiseError(
+            new WorkbenchException(
+              s"Could process StopUpdateMessage for cluster with id ${message.clusterId} because it was not found in the database"
+            )
+          )
+      }
+
+  private def handleClusterTransitionFinished(message: ClusterTransitionFinishedMessage): IO[Unit] =
+    message.clusterFollowupDetails.clusterStatus match {
+      case Stopped => {
+        for {
+          clusterOpt <- dbRef.inTransaction { clusterQuery.getClusterById(message.clusterFollowupDetails.clusterId) }
+          savedMasterMachineType <- dbRef.inTransaction {
+            followupQuery.getFollowupAction(message.clusterFollowupDetails)
+          }
+          result <- clusterOpt match {
+            case Some(resolvedCluster) if resolvedCluster.status != ClusterStatus.Stopped =>
+              IO.raiseError(
+                new WorkbenchException(
+                  s"Unable to process message ${message} for cluster ${resolvedCluster.projectNameString} in status ${resolvedCluster.status.toString}, when the monitor signalled it stopped as it is not stopped."
+                )
+              )
+
+            case Some(resolvedCluster) => {
+
+              savedMasterMachineType match {
+                case Some(machineType) =>
+                  for {
+                    // perform gddao and db updates for new resources
+                    _ <- clusterHelper.updateMasterMachineType(resolvedCluster, MachineType(machineType))
+                    // start cluster
+                    _ <- clusterHelper.internalStartCluster(resolvedCluster)
+                    // clean-up info from follow-up table
+                    _ <- dbRef.inTransaction { followupQuery.delete(message.clusterFollowupDetails) }
+                  } yield ()
+                case None => IO.unit //the database has no record of a follow-up being needed. This is a no-op
+              }
+            }
+
+            case None => IO.raiseError(ClusterNotFoundException(message.clusterFollowupDetails.clusterId))
+          }
+        } yield result
+      }
+
+      //No actions for other statuses yet. There is some logic that will be needed for all other cases (i.e. the 'None' case where no cluster is found in the db and possibly the case that checks for the data in the DB)
+      // TODO: Refactor once there is more than one case
+      case _ => IO.unit
+    }
+
+}
+
+sealed trait LeoPubsubMessage {
```

Comment:
A couple questions about the data model here:
- should the base trait have `clusterId` and `messageType`?
- it's not clear to me why `ClusterFollowupDetails` needs a `ClusterStatus`, since the status is on the cluster record itself

My proposal in code:
```
sealed trait LeoPubsubMessage {
  def messageType: String
  def clusterId: Long
}
object LeoPubsubMessage {
  final case class StopUpdateMessage(updatedMachineConfig: MachineConfig, clusterId: Long) extends LeoPubsubMessage {
    val messageType = "stopUpdate"
  }
  final case class ClusterTransitionFinishedMessage extends LeoPubsubMessage {
    val messageType = "transitionFinished"
  }
}
```

Let me know what you think..

Reply:
My thought behind this was, The transition finished message needs the status due to the asynchronous nature of the message queue and the need to ensure that the status in the DB matches the status in the message before proceeding with the action. 

In other words, if it was a simple 'transition done' message, the user could quickly delete the cluster, or some other unexpected status change could happen (it errors somehow), causing a failure without a clear cause in the log. 

It is true that the check is a bit duplicative with the methods that are called though

Reply:
That makes sense.. the message is published from the ClusterMonitorActor when the cluster is Stopped, but something could have happened in the meantime